### ISSUE TITLE
Proposal: harmonisation of `docid` field to combine `docnumber`, `id`, `docid`

### ISSUE THREAD

**Issue by ronaldtse** [relaton/relaton#100]
Due to the long discussions and resulting confusion and controversy around document identifiers in Relaton data (@strogonoff @opoudjis @andrew2net ), I propose we harmonise all of these identifier "things" under one field.

I would prefer having a docid array that has different types (and maybe scopes) of these identifiers.

* Relaton (@andrew2net)  uses `id` for a unique reference within a dataset. That is actually a machine-readable identifier for generating filenames scoped unique within an authoritative source (publisher).

* Metanorma (@opoudjis ) uses `docid` with `primary: true` for dated references for displaying in citations, and manipulating them when necessary according to the `type` attribute.
* Metanorma (@opoudjis ) now needs an undated PubID for ISO to display on the ISO cover page.
* Metanorma (@opoudjis ) uses `docnumber` as a part of the PubID for some operations. I am unsure where exactly it is used. Most likely, it should be superseded by multiple forms of PubIDs.
* Metanorma IETF (@opoudjis ) uses the `type: IETF; scope: anchor` is used to generate the anchor for xml2rfc.
* BibXML Service (@strogonoff ) uses the `docid` with `primary: true` is used for uniquely identifying a bibliographic item (de-duplication) across all Relaton datasets.

Is this set of use cases accurate?

**Comment by strogonoff** [comment]
One note regarding `docid`:

- I believe, per discussed design, it’s intended to denote an identifier the `id` of which is used by humans when referencing the bibliographic item (unambiguously, globally and uniquely).

**Comment by andrew2net** [comment]
> * Relaton (@andrew2net)  uses `id` for a unique reference within a dataset. That is actually a machine-readable identifier for generating filenames scoped unique within an authoritative source (publisher).

No, I don't use`id`. I use `docid` with `primary: true` and in some cases `docnumber` (for generating filenames)

**Comment by ronaldtse** [comment]
@andrew2net got it, then what is `id` for? No one uses it?

**Comment by strogonoff** [comment]
@andrew2net I was trying to say on the call, if you used `docnumber` for generating filenames then I-D versions would not have unique filenames… so it’s not `docnumber`, apparently

**Comment by andrew2net** [comment]
@ronaldtse I don't know. The `id` is in the grammar, so I've implemented it. Lets ask @opoudjis

**Comment by ronaldtse** [comment]
> so it’s not docnumber, apparently

It's not, it seems that the filenames are constructed from another mechanism. I think the `id` should be a machine-readable ID.

**Comment by ronaldtse** [comment]
I've added the task for ISO and IETF here:
* https://github.com/relaton/relaton-iso/issues/107
* https://github.com/relaton/relaton-ietf/issues/74

This means we also have to do the same for all other document types: IEC, ITU, OGC, CIE, NIST, etc...

**Comment by andrew2net** [comment]
> @andrew2net I was trying to say on the call, if you used `docnumber` for generating filenames then I-D versions would not have unique filenames… so it’s not `docnumber`, apparently

@strogonoff I use `docnumber` for generating filenames in some cases but not for I-D. Sorry for misinformation.

**Comment by ronaldtse** [comment]
While we're at dealing with `docid`, let's also consider the following:
https://github.com/relaton/relaton-nist/commit/e18b5c6adf17f354d0a47a80ea6a66f4debcdb7f#r70684855
```xml
  <docidentifier type="NIST" primary="true">NIST IR 8011-1</docidentifier>
  <docidentifier type="DOI">10.6028/NIST.IR.8011-1</docidentifier>
  <docidentifier type="NIST" scope="anchor">NIST.IR.8011-1</docidentifier>
```

* `type` should be an enumerated value, so we should use lowercase. We also need documentation on what all these `type` values are.
* `scope="anchor"` was originally an IETF BibXML thing -- the use of the "anchor" is what we call the "reference tag" in ISO 690 -- a short handle for referring to a bibliographic item.

**Comment by strogonoff** [comment]
@ronaldtse 

* I think using lowercase for “type” would help make it clear that it’s not publisher org abbreviation. If this changes, we’ll have to check if there are any dependencies on exact case-sensitive values though…
* For `scope="anchor"`, BibXML service now uses it for constructing the `anchor` attribute on XML’s `<reference>`/`<referencegroup>` tags—let me know if we change key name (and if we change the values, then I’ll need to review the logic of how anchors are constructed in the service)

**Comment by ronaldtse** [comment]
@strogonoff Can I confirm that we no longer need `scope="anchor"` in Relaton data. This will be a massive saver in non-IETF origin bibliographic items for @andrew2net .

**Comment by strogonoff** [comment]
@ronaldtse Will the anchor still be provided somewhere in data? In this shape: https://github.com/ietf-ribose/relaton-data-ids/blob/cd11648aa981ac87d781b8dac5598e0d56514940/data/draft-3k1n-6tisch-alice0-01.yaml#L18-L20

If not, let me know in advance, we’ll have to port corresponding these anchor formatting rules from wherever they are defined in Ruby and make them part of BibXML service Python codebase.

**Comment by ronaldtse** [comment]
@strogonoff a `docid` of scope `anchor` will be provided for IETF-sourced data, i.e. RFC, RFC subseries, I-Ds, IANA. No `anchor` will be provided for all other sources.

**Comment by ronaldtse** [comment]
@strogonoff if possible, I prefer to remove `anchor` everywhere because it is not technically part of bibliographic data.

**Comment by strogonoff** [comment]
> @strogonoff if possible, I prefer to remove `anchor` everywhere because it is not technically part of bibliographic data.

@ronaldtse I don’t mind, but:

1. There may be regressions in anchor behavior of BibXML service while the code is ported
2. Didn’t Metanorma use `docid` with `scope=anchor` somehow, too?
3. Generally, our Relaton datasets currently have [many non-standard properties that are not really part of bibliographic data model](https://github.com/ietf-ribose/relaton-data-3gpp/blob/8fd5fb34605420cbc910687fc878cb0716c7a9a7/data/TR_03.49_PH2_4.6.0.yaml#L62-L78). If `anchor` is IETF-specific, it’s just another such case. I imagine we could keep anchors in a similar way. (And as previously suggested Relaton model could normalize custom extensions by moving them under a dedicated key, to avoid polluting the top level of bibliographic item data and possible future clashes between semantics.)

**Comment by ronaldtse** [comment]
> 1. There may be regressions in anchor behavior of BibXML service while the code is ported

Fully understand.

> 2. Didn’t Metanorma use `docid` with `scope=anchor` somehow, too?

I'm not sure, but Metanorma technically does not need to rely on `anchor`. Again, `anchor` has always been an IETF/xml2rfc specific thing.

> 3. Generally, our Relaton datasets currently have [many non-standard properties that are not really part of bibliographic data model](https://github.com/ietf-ribose/relaton-data-3gpp/blob/8fd5fb34605420cbc910687fc878cb0716c7a9a7/data/TR_03.49_PH2_4.6.0.yaml#L62-L78). 

This is not intended -- we have an `ext` element that was used to store flavor-specific information. We probably should use a dedicated key for particular flavor-specific information (ping @andrew2net )

> (And as previously suggested Relaton model could normalize custom extensions by moving them under a dedicated key, to avoid polluting the top level of bibliographic item data and possible future clashes between semantics.)

Agree.

**Comment by strogonoff** [comment]
Understood. I am OK removing `docid` with `scope=anchor`, but perhaps we could move those IDs somewhere else (perhaps under `ext`) instead of getting rid of them completely?
